images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go

```diff
@@ -0,0 +1,214 @@
+			h.logger.Info("VirtualMachineIP address omitted in the spec: allocate the new one")
+			var err error
+			currentIp.Status.Address, err = h.allocateNewIP(state.AllocatedIPs())
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			currentIp.Status.Address = currentIp.Spec.StaticIP
+		}
+
+		if !h.isAvailableAddress(currentIp.Status.Address, state.AllocatedIPs()) {
+			h.logger.Info("VirtualMachineIP cannot be created: the address has already been allocated for another vmip", "address", currentIp.Spec.StaticIP)
+			return reconcile.Result{}, nil
+		}
+
+		leaseName := common.IpToLeaseName(currentIp.Status.Address)
+
+		h.logger.Info("Create lease",
+			"leaseName", leaseName,
+			"refName", state.VirtualMachineIP().Name().Name,
+			"refNamespace", state.VirtualMachineIP().Name().Namespace,
+		)
+
+		err := h.client.Status().Update(ctx, currentIp)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed update status for vmip: %w", err)
+		}
+
+		err = h.client.Create(ctx, &virtv2.VirtualMachineIPAddressLease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: leaseName,
+			},
+			Spec: virtv2.VirtualMachineIPAddressLeaseSpec{
+				VirtualMachineIPAddressRef: &virtv2.VirtualMachineIPAddressLeaseIpAddressRef{
+					Name:      state.VirtualMachineIP().Name().Name,
+					Namespace: state.VirtualMachineIP().Name().Namespace,
+				},
+			},
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	case lease.Status.Phase == "":
+		h.logger.Info("Lease is not ready: waiting for the lease")
+		return reconcile.Result{Requeue: true, RequeueAfter: 2 * time.Second}, nil
+
+	case util.IsBoundLease(lease, state.VirtualMachineIP()):
+		h.logger.Info("Lease already exists, VirtualMachineIP ref is valid")
+		return reconcile.Result{}, nil
+
+	case lease.Status.Phase == virtv2.VirtualMachineIPAddressLeasePhaseBound:
+		h.logger.Info("Lease is bounded to another VirtualMachineIP: recreate VirtualMachineIP when the lease is released")
+		return reconcile.Result{}, nil
+
+	default:
+		h.logger.Info("Lease is released: set binding")
+
+		if lease.Spec.VirtualMachineIPAddressRef.Namespace != state.VirtualMachineIP().Name().Namespace {
+			return reconcile.Result{}, fmt.Errorf("the selected VirtualMachineIP lease belongs to a different namespace: %s", lease.Spec.VirtualMachineIPAddressRef.Namespace)
+		}
+
+		lease.Spec.VirtualMachineIPAddressRef = &virtv2.VirtualMachineIPAddressLeaseIpAddressRef{
+			Name:      state.VirtualMachineIP().Name().Name,
+			Namespace: state.VirtualMachineIP().Name().Namespace,
+		}
+
+		err := h.client.Update(ctx, lease)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		changedIp.Status.Address = common.LeaseNameToIP(lease.Name)
+		return reconcile.Result{}, h.client.Status().Update(ctx, changedIp)
+	}
+}
+
+func (h IPLeaseHandler) isAvailableAddress(address string, allocatedIPs util.AllocatedIPs) bool {
+	ip := net.ParseIP(address)
+
```

Comment:
Consider move isAvailableAddress, allocateNewIP, inc from the handler and create a service.